src/bin/pg_autoctl/keeper.c

```diff
@@ -1317,3 +1302,211 @@ keeper_state_as_json(Keeper *keeper, char *json, int size)
 	/* strlcpy returns how many bytes where necessary */
 	return len < size;
 }
+
+
+/*
+ * keeper_update_group_hba updates updates the HBA file to ensure we have two
+ * entries per other node in the group, allowing for both replication
+ * connections and connections to the --dbname.
+ */
+bool
+keeper_update_group_hba(Keeper *keeper, NodeAddressArray *diffNodesArray)
+{
+	LocalPostgresServer *postgres = &(keeper->postgres);
+	PostgresSetup *postgresSetup = &(postgres->postgresSetup);
+	PGSQL *pgsql = &(postgres->sqlClient);
+
+	char hbaFilePath[MAXPGPATH] = { 0 };
+	char *authMethod = pg_setup_get_auth_method(postgresSetup);
+
+	/* early exit when we're alone in the group */
+	if (diffNodesArray->count == 0)
+	{
+		return true;
+	}
+
+	sformat(hbaFilePath, MAXPGPATH, "%s/pg_hba.conf", postgresSetup->pgdata);
+
+	if (!pghba_ensure_host_rules_exist(hbaFilePath,
+									   diffNodesArray,
+									   postgresSetup->ssl.active,
+									   postgresSetup->dbname,
+									   PG_AUTOCTL_REPLICA_USERNAME,
+									   authMethod))
+	{
+		log_error("Failed to edit HBA file \"%s\" to update rules to current "
+				  "list of nodes registered on the monitor",
+				  hbaFilePath);
+		return false;
+	}
+
+	/*
+	 * Only reload if Postgres is known to be running. If it's not running, we
+	 * edited the HBA and it's going to take effect at next restart of
+	 * Postgres, so we're good here.
+	 */
+	if (pg_setup_is_running(postgresSetup))
+	{
+		if (!pgsql_reload_conf(pgsql))
+		{
+			log_error("Failed to reload the postgres configuration after adding "
+					  "the standby user to pg_hba");
+			return false;
+		}
+	}
+
+	return true;
+}
+
+
+/*
+ * keeper_refresh_other_nodes call pgautofailover.get_other_nodes on the
+ * monitor and refreshes the keeper otherNodes array with fresh information,
+ * including each node's LSN position.
+ *
+ * When forceCacheInvalidation is true, instead of trusting our previous value
+ * for the keeper otherNodes array, keeper_refresh_other_nodes() instead runs
+ * through the whole monitor.get_other_nodes() result and updates HBA rules for
+ * all entries there. That's necessary after a pg_basebackup for instance.
+ * which will copy over the origin's pg_hba.conf.
+ */
+bool
+keeper_refresh_other_nodes(Keeper *keeper, bool forceCacheInvalidation)
+{
+	Monitor *monitor = &(keeper->monitor);
+	KeeperConfig *config = &(keeper->config);
+	PostgresSetup *postgresSetup = &(keeper->postgres.postgresSetup);
+
+	NodeAddressArray *otherNodesArray = &(keeper->otherNodes);
+	NodeAddressArray oldNodesArray = *otherNodesArray;
+	NodeAddressArray diffNodesArray = { 0 };
+
+	char *host = config->nodename;
+	int port = postgresSetup->pgport;
+
+	log_trace("keeper_refresh_other_nodes");
+
+	if (!monitor_get_other_nodes(monitor, host, port, ANY_STATE, otherNodesArray))
+	{
+		log_error("Failed to get_other_nodes() on the monitor");
```

Comment:
Maybe good to also reset `otherNodesArray` here like is done in the other error case. I don't think it's strictly necessary, because of the details of the parsing in `monitor_get_other_nodes`. But I think that's a fairly specific implementation detail, that's best not to be relied upon.

Another way would be to it the other way around and only change `keeper->otherNodes` when we finish successfully.
Instead of `otherNodesArray` you could have something like:
```c
NodeAddressArray newNodesArray = { 0 };
NodeAddressArray oldNodesArray = keeper->otherNodes;
NodeAddressArray diffNodesArray = { 0 };
if (!monitor_get_other_nodes(monitor, host, port, ANY_STATE, newNodesArray)) {
    return false;
}
// do things
keeper->otherNodes = newNodesArray;
return true
```

I think I like this second approach a bit better, since I think it makes the code easier to understand.

Reply:
Fair point, done that way now.